### PR TITLE
Fix build host when only the .net 6 SDK is installed

### PR DESCRIPTION
**Pull request by dibarbet**
Was caught by integration tests
```
[Error - 12:58:15 AM] [BuildHost PID 3252] The BuildHost process exited with -532462766. Process output:
Unhandled exception. System.IO.FileLoadException: Could not load file or assembly 'System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)
File name: 'System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at Microsoft.CodeAnalysis.MSBuild.Program.Main(String[] args)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.CodeAnalysis.MSBuild.Program.Main(String[] args)
   at Microsoft.CodeAnalysis.MSBuild.Program.<Main>(String[] args)
```

SCI was not getting shipped in the output, so when using a .net6 runtime (.net6 SDK only installed, launch server via .exe instead of .dll) it failed to find the 8.0 dependency.
Adding an explicit reference seems to fix the issue

### CODE REVIEW

**Review by jasonmalinowski — APPROVED** [review]
As a workaround this seems fine, but can we file a bug to investigate why this is necessary? This feels like a bug in MSBuild or NuGet...

**Comment by dibarbet** [comment]
> As a workaround this seems fine, but can we file a bug to investigate why this is necessary? This feels like a bug in MSBuild or NuGet...

it kind of seems expected to me?  We didn't directly reference it and we're not shipping self contained, so wouldn't it generally not be added to the output dir?  Might be entirely wrong here.

**Comment by sharwell** [comment]
> We didn't directly reference it and we're not shipping self contained

I was expecting we would have an implicit reference to it from the SDK. From there, the `<PackageVersion>` element would override the version of the implicit reference.

**Review by arkalyanms — APPROVED** [review]

File: src/Tools/PrepareTests/TestDiscovery.cs [hunk 1]
```diff
@@ -85,14 +86,22 @@ static bool RunWorker(string dotnetPath, string pathToWorker, string pathToAssem
...
+        var errorOutput = new StringBuilder();
+
         using (var pipeServer = new AnonymousPipeServerStream(PipeDirection.Out, HandleInheritability.Inheritable))
         {
             // Pass the client process a handle to the server.
             arguments.Add(pipeServer.GetClientHandleAsString());
             pipeClient.StartInfo.Arguments = string.Join(" ", arguments);
             pipeClient.StartInfo.UseShellExecute = false;
+
+            // Errors will be logged to stderr, redirect to us so we can capture it.
```
Comment (dibarbet) [inline src/Tools/PrepareTests/TestDiscovery.cs:98]: errors were getting swallowed so it was just failing with no info.  this ensures error output from the discovery worker is captured

File: src/Tools/TestDiscoveryWorker/Program.cs [hunk 2]
```diff
@@ -24,76 +24,85 @@
     return ExitFailure;
 }
 
-using var pipeClient = new AnonymousPipeClientStream(PipeDirection.In, args[0]);
-using var sr = new StreamReader(pipeClient);
-string? output;
-
-// Wait for 'sync message' from the server.
-do
+try
```
Comment (dibarbet) [inline src/Tools/TestDiscoveryWorker/Program.cs:27]: similar change to ensure exception details are captured if something goes wrong.

File: src/Workspaces/Core/MSBuild.BuildHost/Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.csproj [hunk 3]
```diff
@@ -30,7 +30,9 @@
     <PackageReference Include="Microsoft.Build.Tasks.Core" VersionOverride="17.3.2" ExcludeAssets="Runtime" PrivateAssets="All" />
     <PackageReference Include="Microsoft.Build.Locator" PrivateAssets="All" />
     <PackageReference Include="System.CommandLine" />
+    <PackageReference Include="System.Collections.Immutable" />
     <PackageReference Include="Newtonsoft.Json" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'" />
```
Comment (dibarbet) [inline src/Workspaces/Core/MSBuild.BuildHost/Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.csproj:35]: this is the actual fix for the test discovery issue.  Test discovery was launching the build host - but the .netframework build host failed with a missing reference to this.

I'm not sure exactly why that occurred.

Comment (jasonmalinowski) [reply src/Workspaces/Core/MSBuild.BuildHost/Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.csproj:35]: Why the condition on the TFM? Does having that in the .NET Core side of the build break something?

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**